src/main/scala/cromwell/engine/backend/runtimeattributes/CromwellRuntimeAttributes.scala

```diff
@@ -24,163 +29,187 @@ case class CromwellRuntimeAttributes(docker: Option[String],
+    val attributes = for {
+      keys <- validateKeys(wdlRuntimeAttributes, backendCall.backend.backendType)
+      attributesFromTask <- TryUtil.sequenceMap(wdlRuntimeAttributes.evaluate(backendCall.lookupFunction(backendCall.locallyQualifiedInputs), backendCall.engineFunctions))
+      attributesWithDefaults <- Try(getAttributesWithDefaults(attributesFromTask, workflowOptions))
+      validatedAttributes <- validateRuntimeAttributes(attributesWithDefaults)
+    } yield validatedAttributes
 
-  def apply(wdlRuntimeAttributes: RuntimeAttributes, workflowOptions: WorkflowOptions, backendType: BackendType) = {
-    val attributesMap = AttributeMap(wdlRuntimeAttributes.attrs)
 
-    val withDefaults = new AttributeMapTrait {
-      override def keys = attributesMap.keys ++ workflowOptions.getDefaultRuntimeOptionKeys
-      override def unsupportedKeys(bt: BackendType): Seq[String] = attributesMap.unsupportedKeys(bt)
-      override def getSeq(key: RuntimeKey): Option[Seq[String]] = attributesMap.getSeq(key)
-      override def get(key: RuntimeKey): Option[String] = attributesMap.get(key) match {
-        case Some(x) => Some(x)
-        case None => workflowOptions.getDefaultRuntimeOption(key.key).toOption
-      }
+    attributes.get
+  }
+
+  private def getAttributesWithDefaults(taskAttributes: Map[String, WdlValue], workflowOptions: Option[WorkflowOptions]): Map[String, WdlValue] = {
+    val defaultAttributes = workflowOptions match {
+      case Some(options) => defaultValues(options)
+      case None => defaultValues
     }
 
-    fromAttributeMap(withDefaults, backendType)
+    // Merge the two maps, prefer values from the latter
+    defaultAttributes ++ taskAttributes
   }
 
-  private def fromAttributeMap(attributeMap: AttributeMapTrait, backendType: BackendType): CromwellRuntimeAttributes = {
+  def validateKeys(wdlRuntimeAttributes: RuntimeAttributes, backendType: BackendType): Try[RuntimeAttributes] = {
     /**
       *  Warn if keys are found that are unsupported on this backend.  This is not necessarily an error, at this point
       *  the keys are known to be supported on at least one backend other than this.
       */
-    attributeMap.unsupportedKeys(backendType).foreach(unsupportedKeyMessage => log.info(unsupportedKeyMessage))
-
-    val attributeMapNel = validateRequiredKeys(attributeMap, backendType)
-    val runtimeAttributeNel = validateRuntimeAttributes(attributeMap)
-    val validatedRuntimeAttributes = (attributeMapNel |@| runtimeAttributeNel) { (_, r) => r }
-
-    validatedRuntimeAttributes match {
-      case Success(r) => r
-      case Failure(f) =>
-        throw new IllegalArgumentException() with ThrowableWithErrors {
-          val message = "RuntimeAttribute is not valid."
-          val errors = f
-        }
+    unsupportedKeys(wdlRuntimeAttributes.attrs.keys, backendType) foreach log.warn
```

Comment:
The reason why it was like this is that some tests calls `unsupportedKeys`, but maybe if I can figure out how to make assertions on log messages (outside of Akka), I can change it to your suggestion